federation/pkg/federation-controller/sync/controller.go

```diff
@@ -320,113 +323,90 @@ func (s *FederationSyncController) reconcile(namespacedName types.NamespacedName
-		kind, namespacedName)
-	// Add the required finalizers before creating the resource in underlying clusters.
+	glog.V(3).Infof("Ensuring finalizers for %s %q", kind, key)
 	obj, err = s.deletionHelper.EnsureFinalizers(obj)
 	if err != nil {
-		glog.Errorf("failed to ensure delete object from underlying clusters finalizer in %s %s: %v",
-			kind, namespacedName, err)
+		glog.Errorf("Failed to ensure finalizers for %s %q: %v", kind, key, err)
 		return statusError
 	}
 
-	glog.V(3).Infof("Syncing %s %s in underlying clusters", kind, namespacedName)
+	glog.V(3).Infof("Syncing %s %q in underlying clusters", kind, key)
 
 	clusters, err := s.informer.GetReadyClusters()
 	if err != nil {
-		glog.Errorf("failed to get cluster list: %v", err)
+		glog.Errorf("Failed to get cluster list: %v", err)
 		return statusNotSynced
 	}
 
-	operations := make([]util.FederatedOperation, 0)
-	for _, cluster := range clusters {
-		clusterObj, found, err := s.informer.GetTargetStore().GetByKey(cluster.Name, key)
-		if err != nil {
-			glog.Errorf("failed to get %s from %s: %v", key, cluster.Name, err)
-			return statusError
-		}
-
-		// The data should not be modified.
-		desiredObj := s.adapter.Copy(obj)
-
-		if !found {
-			operations = append(operations, util.FederatedOperation{
-				Type:        util.OperationTypeAdd,
-				Obj:         desiredObj,
-				ClusterName: cluster.Name,
-				Key:         key,
-			})
-		} else {
-			clusterObj := clusterObj.(pkgruntime.Object)
-
-			// Update existing resource, if needed.
-			if !s.adapter.Equivalent(desiredObj, clusterObj) {
-				operations = append(operations, util.FederatedOperation{
-					Type:        util.OperationTypeUpdate,
-					Obj:         desiredObj,
-					ClusterName: cluster.Name,
-					Key:         key,
-				})
-			}
-		}
+	operations, err := clusterOperations(s.adapter, clusters, obj, key, func(clusterName string) (interface{}, bool, error) {
+		return s.informer.GetTargetStore().GetByKey(clusterName, key)
+	})
+	if err != nil {
+		glog.Error(err)
+		return statusError
 	}
-
 	if len(operations) == 0 {
-		// Everything is in order
 		return statusAllOK
 	}
 
 	err = s.updater.Update(operations)
 	if err != nil {
-		glog.Errorf("failed to execute updates for %s: %v", key, err)
+		glog.Errorf("Failed to execute updates for %s %q: %v", kind, key, err)
 		return statusError
 	}
 
 	// Evertyhing is in order but let's be double sure
 	return statusNeedsRecheck
 }
 
+func (s *FederationSyncController) objFromCache(kind, key string) (pkgruntime.Object, error) {
+	cachedObj, exist, err := s.store.GetByKey(key)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to query %s store for %q: %v", kind, key, err)
+	}
+	if !exist {
+		return nil, nil
+	}
+
+	// Create a copy before modifying the resource to prevent racing with other readers.
+	copiedObj, err := api.Scheme.DeepCopy(cachedObj)
+	if err != nil {
+		return nil, fmt.Errorf("Error in retrieving %s %q from store: %v", kind, key, err)
```

Comment:
That's not the case here, though, the only caller is just glog'ing the error directly.